### DISCUSSION TITLE
A member with the name 'name' already exists on 'OpenAI.Chat.Message'

### DISCUSSION THREAD

**Discussion by MegaMax5000** [RageAgainstThePixel/com.openai.unity#171] [Q&A] [answered]
# Bug Report
The following error message occurs when attempting ChatEndpoint.GetCompletionAsync():

JsonSerializationException: A member with the name 'name' already exists on 'OpenAI.Chat.Message'. Use the JsonPropertyAttribute to specify another name.

I apologize if this is not the proper space to bring this up, I could not start a discussion in the Q&A section of the Discussions tab. 

## Overview
Upon looking through the source I see that in Message.cs there is a serialized field "name" as well as using the JsonPropertyAttribute to specify the public "Name" field to serialize as "name".  This might be causing the issue. 

## To Reproduce
The following code reproduces this problem on the "await api.ChatEndpoint.GetCompletionAsync(chatRequest);" line.

```csharp
private async Task<string> SendMessage(string message)
{
    var api = new OpenAIClient(GPT_API_KEY);

    var messages = new List<Message>
    {
        new Message(Role.System, GPT_SYSTEM_ROLE),
        new Message(Role.User, message)
    };
    var chatRequest = new ChatRequest(messages, GPT_MODEL_ID);
    var response = await api.ChatEndpoint.GetCompletionAsync(chatRequest);
    var choice = response.FirstChoice;

    return choice.Message.Content.ToString();
}
```

## Expected behavior
I would expect that error to not occur and for the GetCompletionAsync() to complete as expected.

## Additional context
I am using this package in Unity 2022.3.10f1.

**Comment by MegaMax5000** [comment]
Worth noting when I modified the local package cache version of Message.cs to test a fix, I got the same problem for "role" and "content" and some others once "name" was fixed.

**Comment by StephenHodgson** [comment]
> Worth noting when I modified the local package cache version of Message.cs to test a fix, I got the same problem for "role" and "content" and some others once "name" was fixed.

Yeah this is strange, did you somehow modify the json serialization settings? This should not be possible.

**Comment by MegaMax5000** [comment]
My project is using Newtonsoft Json, could that be causing collisions?

**Comment by StephenHodgson** [comment]
Test installing this in a new project with the same editor version. If it works, then you have another package that is modifying the serialization settings.

**Reply by MegaMax5000** [reply]
Tried in a new empty project and the same script works as expected. Where/how would the json serialization settings be viewed and changed in my other project?

**Reply by MegaMax5000** [reply]
It does look like removing [SerializeField] from the private fields already being serialized with public fields with [JsonProperty] attributes on my local cached version as a test did clear up the errors without tracking down or changing json serialization settings. 

The fields: name, role, content

More info found here: https://forum.unity.com/threads/newtonsoft-json-serialization-works-but-deserialization-doesnt.1289264/

Of course, modifying local caches is just a temporary debug check and not an actual solution.

**Reply by StephenHodgson** [reply]
You need to figure out what other package is modifying the json serialization settings

**Reply by StephenHodgson** [reply]
I think someone else also ran into this before, their fix was to remove/edit the settings in the offending package.

**Reply by MegaMax5000** [reply, marked as answer]
It looks like this is the package in question in case it's helpful for future reference: https://github.com/applejag/Newtonsoft.Json-for-Unity.Converters/

OpenAI stuff works when I modify some of those settings.

**Reply by StephenHodgson** [reply]
Yep that was the one. 

A particularly don't like how it changes serialization settings for classes that it shouldn't be modifying.
